### ISSUE TITLE
What does `T` mean here?

### ISSUE THREAD

**Issue by CoinCheung** [qqwweee/keras-yolo3#322]
Hi, 

I am trying to understand your code, and when I read to this line: 
https://github.com/qqwweee/keras-yolo3/blob/e6598d13c703029b2686bc2eb8d5c09badf42992/yolo3/model.py#L237

I do not understand what the `m` and `T` mean here. What do they mean?

By the way, I noticed that, in the paper, the author mentioned that he assigned only one anchor to each gt bbox, is this the case of your implementation? If I did not misunderstand the paper, we should assign the anchor with the best iou to a gt, is this correct?

**Comment by ancientghost** [comment]
For your first question:

m is the batch size and T should be __max_boxes__ (default is 20), which can be modified in __get_random_data__ function.

**Comment by CoinCheung** [comment]
I saw in the paper, that the classification logits are past through a sigmoid activation, and then put into a classification. But it seems that in this implementation, `keras.binary_crossentropy` is directly used in classification?

**Comment by CoinCheung** [comment]
@ancientghost Thanks a lot. One more question please: 
In yolov3, there are three layers that are used to generate predictions at scale of 8, 16, and 32. There are 9 sorts of anchors, and they are divided into 3 groups and used in these three layers. Then given a gt, the paper said that only one anchor should be assigned to be positive, does this mean one positive anchor at each layer which makes 3 actual positive anchors for each gt, or is there only one positive anchor from all the three layers, which makes the other 8 anchors to be negative or ignored?

**Comment by ancientghost** [comment]
@CoinCheung There should be only one positive anchor for each GT from all the three layers.

**Comment by CoinCheung** [comment]
@ancientghost  Thanks. And when I divide positive and negative matches, do I need to compute the iou between the anchors and gts , or between offset-adjusted anchors and gts? I noticed that the anchors can be adjusted to bounding box with the output of the model, do I need to use these bounding boxes or simply use the anchors? 

If we need simply use anchors, does it mean the regression targets for one certain image are fixed values, since the gts and anchors are all fixed values?

**Comment by ancientghost** [comment]
@CoinCheung Actually you need to calculate both, but in different part.

First, you need to assign you target. That concerns which anchor to use (you have 9 anchor boxes in 3 scales). According to the paper, you need to select one most similar anchor with the gt. But how to evaluate the similarity? That is iou. So, in this step you need calculate the iou between the anchors and gts (check __preprocess_true_boxes__ function). __Note:__ In this step, you don't need to xy information. Because what you need is one more similar anchor box, in other words, you just need to concern the anchors' shape. And furthermore, the anchors don't have xy.  (-_-) I write this because someone confuses about it.

And for the iou between offset-adjusted anchors (this should be your predicted boxes) and gts, of course you need, for the loss (check __box_iou__ function) or anything else. 

I hope i explain it clearly :)

**Comment by CoinCheung** [comment]
@ancientghost Many thanks, just to make sure I have understand this algorithm properly.

When I compute the loss of the bbox shape(width and height), I will never need to consider the position, and I only need to consider the offset between the positive anchor and the gt bbox. Likewise, when I compute the position loss, I do not need to care about the anchor shape either, and I only need to consider the offset between the center of the grid where the anchor lies in and the gt bbox center. Is that true ?

Thus the whole procedure is like this:  First I compute the iou between the 9 anchors and each gt only according to their shape without considering the position, and then I choose a best one as the positive anchor, and neglect the anchors with iou higher than 0.5, leaving the rest as negative anchors. The positive anchor is required in the loss of position (x and y) and classification and shape(width and height) as well as the confidence. The negative anchors are only needed in the loss of confidence, while the neglected anchors make no contribution to the loss.

Please correct me if I have ill understanding to these details.
